<html>
  <head>
    <meta charset="utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>异常处理 | CheckChen</title>
<meta name="description" content="吾日三省吾身" />
<link rel="shortcut icon" href="https://cxp-twx.github.io/favicon.ico?v=1576205079785">
<link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.2/css/all.css" integrity="sha384-fnmOCqbTlWIlj8LyTjo7mOUStjsKC4pOpQbqyi7RrhN7udi9RwhKkMHpvLbHG9Sr" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/katex.min.css">
<link rel="stylesheet" href="https://cxp-twx.github.io/styles/main.css">

<script src="https://cdn.bootcss.com/highlight.js/9.12.0/highlight.min.js"></script>
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Droid+Serif:400,700">



  </head>
  <body>
    <div class="main">
      <div class="main-content">
        <div class="site-header">
  <a href="https://cxp-twx.github.io">
  <img class="avatar" src="https://cxp-twx.github.io/images/avatar.png?v=1576205079785" alt="">
  </a>
  <h1 class="site-title">
    CheckChen
  </h1>
  <p class="site-description">
    吾日三省吾身
  </p>
  <div class="menu-container">
    
      
        <a href="/" class="menu">
          首页
        </a>
      
    
      
        <a href="/archives" class="menu">
          归档
        </a>
      
    
      
        <a href="/tags" class="menu">
          标签
        </a>
      
    
      
        <a href="/post/about" class="menu">
          关于
        </a>
      
    
  </div>
  <div class="social-container">
    
      
    
      
    
      
    
      
    
      
    
  </div>
</div>

        <div class="post-detail">
          <article class="post">
            <h2 class="post-title">
              异常处理
            </h2>
            <div class="post-info">
              <span>
                2018-12-13
              </span>
              <span>
                9 min read
              </span>
              
            </div>
            
            <div class="post-content-wrapper">
              <div class="post-content">
                <h1 id="异常处理">异常处理</h1>
<h3 id="什么异常">什么异常？</h3>
<p>在生活中：平时上下班都平常，今天上班的时候，路上有一起车祸，导致你上班迟到。</p>
<p>所谓异常：就出现了出乎意料的事情。</p>
<p>在代码里：所谓异常也是程序在编写或者运行过程中出现了出乎意料的事情。</p>
<p>比如：你做3D彩票系统，你提示别人输入数字，结果别人输入的还是字符串，出现了crash。</p>
<p>再比如：你把元素放到ArrayList中，但是你给的下标超出了元素的个数，程序也会出现异常。</p>
<h3 id="异常的分类">异常的分类</h3>
<p>在Java中异常一共分为2大类：Error、Exception。</p>
<p>早整个异常体系了，有如下继承关系</p>
<p>Throwable 它是所有异常和错误的父类。</p>
<p>​		Error：它是Throwable的子类。 代表错误，一般指的是系统错误，或者虚拟机错误。错误例如：内存不足，JVM出现问题，也属于错误。错误往往不是程序员能解决的，所以无需解决。</p>
<p>​		Exception：也是Throwable的子类。 代表异常，异常分为2类，运行期异常，另外一类叫非运行期异常。</p>
<p>​				RunTimeException：运行期异常。编译期并不会报错，运行的时候才会报错。</p>
<p>​				其他：非运行期异常。编译期就报错。错误不处理，无法运行程序。</p>
<h4 id="error的示例">Error的示例</h4>
<pre><code class="language-java">public static void main(String[] args) throws FileNotFoundException {
		long[] arr = new long[1024*1024*100*4];//内存不足的错误
	}
}

</code></pre>
<h4 id="exception的示例">Exception的示例</h4>
<h5 id="runtimeexception示例">RunTimeException示例</h5>
<pre><code class="language-java">public class Test {

	public static void main(String[] args) throws FileNotFoundException {
		
		int[] arr1 = {1,2,3,4};
		System.out.println(arr1[4]);//运行期异常， 数组下标越界
		
		int m = 100;
		int n = 0;
		System.out.println(m / n );//运行期异常， 除数为0的异常
	}

}
</code></pre>
<h5 id="非运行期异常的示例">非运行期异常的示例</h5>
<pre><code class="language-java">public class Test {

	public static void main(String[] args) throws FileNotFoundException {	
		FileInputStream fis = new FileInputStream(&quot;C:\\Users\\neal\\Desktop\\a.docx&quot;);//非运行期异常，文件没有找到异常
	}
}
</code></pre>
<h3 id="异常的处理">异常的处理</h3>
<p><strong>如果程序出现了异常，但是并没有对异常进行处理，虚拟机就会处理异常</strong>。虚拟机的处理方式就是结束出现异常的线程。在多线程的环境下，如果A线程出现了异常，只是结束A线程的运行，B，C等线程都能正常的执行。</p>
<p>异常的处理有2种方式：</p>
<p>方式一：try ...catch...finally  直接处理异常</p>
<p>方式二：thows	抛出异常，交给这个方法的调用者处理。</p>
<blockquote>
<p>如果有一个a方法，a方法内部可能出现异常，但是a方法内部通过thows抛出了异常，相当于告诉调用a方法的那段代码，a方法有可能有异常，但是我没有处理，交给调用a方法的代码去处理。</p>
</blockquote>
<h4 id="trycatchfinally">try...catch...finally</h4>
<p>把有可能出现异常的代码，放到try代码块中。try的作用是试图找到和发现异常。一旦try代码块中有异常，会执行对应的catch代码块。</p>
<pre><code class="language-java">public class Test {

	public static void main(String[] args) {
		
		int m = 100;
		int n = 0;
		int result = 0;
		int[] arr = {1,2,3};
		try {
			result = m / n;
			int num = arr[3];
		}catch(ArithmeticException e) {
//			System.out.println(e.getMessage());
//			System.out.println(e.getLocalizedMessage());
//			System.out.println(e);
			n = 1;
			result = m / n;
		}catch(ArrayIndexOutOfBoundsException e) {
			int num = arr[0];
			System.out.println(num);
		}
		
		System.out.println(result);
	}
}
</code></pre>
<blockquote>
<p>try中一旦发现异常，就会立即跳转到对应的catch块中进行异常处理。try中没有执行到的代码，不再执行。换句话说，try只能捕获第一个异常。</p>
<p>如果有多个异常，写多个try..catch..finally代码段。</p>
<p>多个catch是可以合并的。只需要把catch中的异常范围写的更大即可。</p>
</blockquote>
<p>因为try一旦发现异常，就立即进入了catch块，导致try中的部分代码没法执行。异常处理提供了finally来解决这个问题。</p>
<p>finally也是一个代码块，无论try有没有异常，finally中的代码都会执行。</p>
<p>情形一：try中没有出现异常，catch不执行，finally执行。</p>
<p>情形二：try中出现异常，执行对应的catch，catch执行以后，执行finally。</p>
<p>所以可以把必须要执行的代码写在finally里面。</p>
<pre><code class="language-java">public class Test {

	public static void main(String[] args) {
		
		int m = 100;
		int n = 10;
		int result = 0;
		int[] arr = {1,2,3};
		try {
			result = m / n;
			int num = arr[2];
		}catch(RuntimeException e) {
//			System.out.println(e.getMessage());
//			System.out.println(e.getLocalizedMessage());
//			System.out.println(e);
			n = 1;
			result = m / n;
		}finally{
			System.out.println(&quot;我是finally&quot;);
		}
		
		System.out.println(result);
	}
}
</code></pre>
<h3 id="小节">小节</h3>
<p>Java中异常分为Error和Exception两类</p>
<p>Exception分为RuntimeException和非RuntimeException。</p>
<p>try..catch..finally可以处理异常。</p>
<p>​	try中写，可能出现的异常。</p>
<p>​	catch中写，如何处理异常。异常是可以分开捕获，异常还可以合起来捕获。</p>
<p>​	finally中写那些一定要执行的代码。</p>
<h4 id="throws">throws</h4>
<p>throws是第二种异常处理的方式。这种处理方式，相当于自己没有做出处理，把处理的事情交给了它的上一级（调用者）来处理。</p>
<h4 id="为什么要使用throws">为什么要使用throws？</h4>
<p>一般调用者更清楚异常出来以后应该做什么事情。</p>
<h4 id="throws如何使用">throws如何使用？</h4>
<p>当有异常存在的时候，如果这个异常不适合你处理，适合你的上层（调用者）处理，你应该把这个异常处理的权利交给你的上层。上层就会得到这个异常，它有2种选择：第一，通过try..catch..finally处理异常，第二，继续上抛，抛给它的调用者。</p>
<pre><code class="language-java">public class TestException {

	public static void main(String[] args) throws FileNotFoundException {
		
		//readFile(&quot;C:\\Users\\neal\\Desktop\\微信小程序.md&quot;);
//		try {
//			readFile2(&quot;C:\\Users\\neal\\Desktop\\微信小程序.md&quot;);
//		} catch (FileNotFoundException e) {
//			
//		}
		
		readFile2(&quot;C:\\Users\\neal\\Desktop\\微信小程序.md&quot;);
		
	}
	
	public static void readFile2(String filePath) throws FileNotFoundException {
		FileInputStream fis = new FileInputStream(filePath);
		System.out.println(fis);
	}
	
	public static void readFile(String filePath) {
		try {
			FileInputStream fis = new FileInputStream(filePath);
		} catch (FileNotFoundException e) {
			System.out.println(&quot;打开文件失败，文件不存在&quot;);
			e.printStackTrace();
		}
		
	}
}
</code></pre>
<h4 id="throws的特点">throws的特点</h4>
<p>throws本质上是在做一个声明，它在告诉调用者，我内部的代码可能有某种异常产生。用于警醒调用者。</p>
<p>throws出现在方法参数列表的后面，throws 后面跟着的是具体的异常的类名。</p>
<p>补充一句，throws 后面的异常可以是多个，多个异常之间用逗号隔开。</p>
<p>再补充一句，throws后面的异常可以合并为父异常。</p>
<h3 id="throw">throw</h3>
<p>throw是个动词，它的作用是抛出一个具体的异常。</p>
<p>用法：throw 异常对象；</p>
<blockquote>
<p>你平时代码里出错时控制台打印的这些异常信息，都是系统API里提前写了throw e的原因。</p>
</blockquote>
<h4 id="throw-和throws区别">throw 和throws区别</h4>
<ol>
<li>throw是用来抛出异常的。后面跟的是一个具体的异常对象。 throws是用来声明方法内可能会有异常。</li>
<li>throw写在方法内，throws写在方法的参数列表后面。</li>
</ol>
<h3 id="自定义异常">自定义异常</h3>
<p>虽然系统已经提供了很多写好的异常类，但是由于项目不同，需求不同等，所以可能出现的异常也不同（可能你出现的异常，系统之前没有定义过这样的类）。为了解决这个问题，系统提供了自定义异常的语法。</p>
<h4 id="如何自定义异常">如何自定义异常</h4>
<p>创建一个类，继承于某个异常类即可。</p>
<h5 id="自定义一个运行期异常">自定义一个运行期异常：</h5>
<pre><code class="language-java">public class ChooseOutOfBoundsException extends RuntimeException {

	private static final long serialVersionUID = 1L;

	public ChooseOutOfBoundsException(String message) {
		super(message);
	}

}
</code></pre>
<p>自定义异常的使用：</p>
<pre><code class="language-java">public class Test {
	
	public static void main(String[] args) {
		System.out.println(&quot;请输入您的选择（1-7）：&quot;);
		int n = getChoose();
		System.out.println(n);
	}
	
	public static int getChoose() {
		Scanner sc = new Scanner(System.in);
		int choose = sc.nextInt();
		if(choose &gt; 7 || choose &lt; 1) {
			ChooseOutOfBoundsException e = new ChooseOutOfBoundsException(&quot;您选择的数字超出了范围[1,7]&quot;);
			throw e;
		}
		return choose;
	}
}
</code></pre>
<h5 id="自定义一个非运行期异常">自定义一个非运行期异常</h5>
<pre><code class="language-java">public class ChooseOutOfBoundsException2 extends Exception {
	private static final long serialVersionUID = 1L;

	public ChooseOutOfBoundsException2(String message) {
		super(message);
	}
	
}
</code></pre>
<p>使用非运行期异常</p>
<pre><code class="language-java">public class Test2 {

	public static void main(String[] args) {
		System.out.println(&quot;请输入1-7的数：&quot;);
		int n = 0;
		try {
			n = getChoose();
		} catch (ChooseOutOfBoundsException2 e) {
			System.out.println(&quot;你真调皮&quot;);
			//e.printStackTrace();
		}
		System.out.println(n);
	}
	
	public static int getChoose() throws ChooseOutOfBoundsException2 {
		Scanner sc = new Scanner(System.in);
		int choose = sc.nextInt();
		if(choose &gt; 7 || choose &lt; 1) {
			ChooseOutOfBoundsException2 e = new ChooseOutOfBoundsException2(&quot;选择超出范围[1,7]&quot;);
			throw e;
		}
		return choose;
	}

}
</code></pre>
<figure data-type="image" tabindex="1"><img src="https://img-blog.csdnimg.cn/20191022211334228.jpg?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L0NoZW5fbmVhbF9KYXZh,size_16,color_FFFFFF,t_70" alt="在这里插入图片描述"></figure>

              </div>
              <div class="toc-container">
                <ul class="markdownIt-TOC">
<li><a href="#%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86">异常处理</a><br>
*
<ul>
<li><a href="#%E4%BB%80%E4%B9%88%E5%BC%82%E5%B8%B8">什么异常？</a></li>
<li><a href="#%E5%BC%82%E5%B8%B8%E7%9A%84%E5%88%86%E7%B1%BB">异常的分类</a>
<ul>
<li><a href="#error%E7%9A%84%E7%A4%BA%E4%BE%8B">Error的示例</a></li>
<li><a href="#exception%E7%9A%84%E7%A4%BA%E4%BE%8B">Exception的示例</a>
<ul>
<li><a href="#runtimeexception%E7%A4%BA%E4%BE%8B">RunTimeException示例</a></li>
<li><a href="#%E9%9D%9E%E8%BF%90%E8%A1%8C%E6%9C%9F%E5%BC%82%E5%B8%B8%E7%9A%84%E7%A4%BA%E4%BE%8B">非运行期异常的示例</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#%E5%BC%82%E5%B8%B8%E7%9A%84%E5%A4%84%E7%90%86">异常的处理</a>
<ul>
<li><a href="#trycatchfinally">try...catch...finally</a></li>
</ul>
</li>
<li><a href="#%E5%B0%8F%E8%8A%82">小节</a>
<ul>
<li><a href="#throws">throws</a></li>
<li><a href="#%E4%B8%BA%E4%BB%80%E4%B9%88%E8%A6%81%E4%BD%BF%E7%94%A8throws">为什么要使用throws？</a></li>
<li><a href="#throws%E5%A6%82%E4%BD%95%E4%BD%BF%E7%94%A8">throws如何使用？</a></li>
<li><a href="#throws%E7%9A%84%E7%89%B9%E7%82%B9">throws的特点</a></li>
</ul>
</li>
<li><a href="#throw">throw</a>
<ul>
<li><a href="#throw-%E5%92%8Cthrows%E5%8C%BA%E5%88%AB">throw 和throws区别</a></li>
</ul>
</li>
<li><a href="#%E8%87%AA%E5%AE%9A%E4%B9%89%E5%BC%82%E5%B8%B8">自定义异常</a>
<ul>
<li><a href="#%E5%A6%82%E4%BD%95%E8%87%AA%E5%AE%9A%E4%B9%89%E5%BC%82%E5%B8%B8">如何自定义异常</a>
<ul>
<li><a href="#%E8%87%AA%E5%AE%9A%E4%B9%89%E4%B8%80%E4%B8%AA%E8%BF%90%E8%A1%8C%E6%9C%9F%E5%BC%82%E5%B8%B8">自定义一个运行期异常：</a></li>
<li><a href="#%E8%87%AA%E5%AE%9A%E4%B9%89%E4%B8%80%E4%B8%AA%E9%9D%9E%E8%BF%90%E8%A1%8C%E6%9C%9F%E5%BC%82%E5%B8%B8">自定义一个非运行期异常</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

              </div>
            </div>
          </article>
        </div>

        
          <div class="next-post">
            <div class="next">下一篇</div>
            <a href="https://cxp-twx.github.io/post/spring-ji-lu">
              <h3 class="post-title">
                Spring 记录
              </h3>
            </a>
          </div>
        

        
          
            <link rel="stylesheet" href="https://unpkg.com/gitalk/dist/gitalk.css">
<script src="https://unpkg.com/gitalk/dist/gitalk.min.js"></script>

<div id="gitalk-container"></div>

<script>

  var gitalk = new Gitalk({
    clientID: '3818ab196008bdc444d2',
    clientSecret: '25a6a1be783a52e1a00f165dd3c08fa28fba507a',
    repo: 'cxp-twx.github.io',
    owner: 'cxp-twx',
    admin: ['cxp-twx'],
    id: (location.pathname).substring(0, 49),      // Ensure uniqueness and length less than 50
    distractionFreeMode: false  // Facebook-like distraction free mode
  })

  gitalk.render('gitalk-container')

</script>

          

          
        

        <div class="site-footer">
  Powered by <a href="https://github.com/getgridea/gridea" target="_blank">Gridea</a> | 
  <a class="rss" href="https://cxp-twx.github.io/atom.xml" target="_blank">RSS</a>
</div>

<script>
  hljs.initHighlightingOnLoad()

  let mainNavLinks = document.querySelectorAll(".markdownIt-TOC a");

  // This should probably be throttled.
  // Especially because it triggers during smooth scrolling.
  // https://lodash.com/docs/4.17.10#throttle
  // You could do like...
  // window.addEventListener("scroll", () => {
  //    _.throttle(doThatStuff, 100);
  // });
  // Only not doing it here to keep this Pen dependency-free.

  window.addEventListener("scroll", event => {
    let fromTop = window.scrollY;

    mainNavLinks.forEach((link, index) => {
      let section = document.getElementById(decodeURI(link.hash).substring(1));
      let nextSection = null
      if (mainNavLinks[index + 1]) {
        nextSection = document.getElementById(decodeURI(mainNavLinks[index + 1].hash).substring(1));
      }
      if (section.offsetTop <= fromTop) {
        if (nextSection) {
          if (nextSection.offsetTop > fromTop) {
            link.classList.add("current");
          } else {
            link.classList.remove("current");    
          }
        } else {
          link.classList.add("current");
        }
      } else {
        link.classList.remove("current");
      }
    });
  });

</script>

      </div>
    </div>
  </body>
</html>
